app/(dashboard)/dashboard/interviews/_actions/export.ts

```diff
@@ -0,0 +1,147 @@
+'use server';
+
+import FileExportManager from '~/lib/network-exporters/FileExportManager';
+import { api } from '~/trpc/server';
+import { getServerSession } from '~/utils/auth';
+import { formatExportableSessions, getRemoteProtocolID } from './utils';
+import { trackEvent } from '~/analytics/utils';
+import { type ExportOptions } from '../_components/ExportInterviewsDialog';
+
+type UploadData = {
+  key: string;
+  url: string;
+  name: string;
+  size: number;
+};
+
+type UpdateItems = {
+  statusText: string;
+  progress: number;
+};
+
+type FailResult = {
+  data: null;
+  error: unknown;
+  message: string;
+};
+
+type SuccessResult = {
+  data: UploadData;
+  error: null;
+  message: string;
+};
+
+export const exportSessions = async (
+  interviewIds: { id: string }[],
+  exportOptions?: ExportOptions,
+) => {
+  const session = await getServerSession();
+
+  if (!session) {
+    throw new Error('You must be logged in to export interview sessions!.');
+  }
+
+  const installedProtocols = await api.protocol.get.all.query();
+  const interviewsSessions =
+    await api.interview.get.manyByIds.query(interviewIds);
+
+  const fileExportManager = new FileExportManager(exportOptions);
+
+  fileExportManager.on('begin', () => {
+    // eslint-disable-next-line no-console
+    console.log({
+      statusText: 'Starting export...',
+      percentProgress: 0,
+    });
+  });
+
+  fileExportManager.on('update', ({ statusText, progress }: UpdateItems) => {
+    // eslint-disable-next-line no-console
+    console.log({
+      statusText,
+      percentProgress: progress,
+    });
+  });
+
+  fileExportManager.on('session-exported', (sessionId: unknown) => {
+    if (!sessionId || typeof sessionId !== 'string') {
+      // eslint-disable-next-line no-console
+      console.warn('session-exported event did not contain a sessionID');
+      return;
+    }
+    // eslint-disable-next-line no-console
+    console.log('session-exported success sessionId:', sessionId);
+  });
+
+  fileExportManager.on('error', (errResult: FailResult) => {
+    // eslint-disable-next-line no-console
+    console.log('Session export failed, Error:', errResult.message);
+  });
+
+  fileExportManager.on('finished', ({ statusText, progress }: UpdateItems) => {
+    // eslint-disable-next-line no-console
+    console.log({
+      statusText,
+      percentProgress: progress,
+    });
+  });
+
+  const formattedSessions = formatExportableSessions(
+    interviewsSessions,
+    installedProtocols,
+  );
+
+  // The protocol object needs to be reformatted so that it is keyed by
+  // the sha of protocol.name, since this is what network-exporters use.
+  const reformatedProtocols = Object.values(installedProtocols).reduce(
+    (acc, protocol) => ({
+      ...acc,
+      [getRemoteProtocolID(protocol.name)]: protocol,
+    }),
+    {},
+  );
+
+  const exportJob = fileExportManager.exportSessions(
+    formattedSessions,
+    reformatedProtocols,
```

Comment:
typo: `reformated` -> `reformatted`. Should also just be `formattedProtocols`.

Build `formattedProtocols` in the tRPC endpoint by mapping over the sessions and adding the protocol to a new `Set`. Return this from the endpoint along with the sessions.

Reply:
Done ✅ I'm building the `formattedProtocols` inside the server action rather than the tRPC endpoint because of its cashing behavior, The cashing behavior is useful to retrieve data from DB, so I kept it inside the server action.